### PR TITLE
[WIP] Update LLVM

### PR DESCRIPTION
**Pull request by GeorgeLyon**
There are still some failures in the tests I haven't been able to fix, I would appreciate it if someone was able to take a look.

### CODE REVIEW

**Comment by lattner** [comment]
Patch looks good, what are the failures you're seeing?

**Comment by GeorgeLyon** [comment]
Whoops, I though this would be apparent in the CI, but it looks like the error in the CI is that those commits aren't yet in `circt/llvm` (maybe we should point to `llvm/llvm-project` directly instead? :-) 

I've submitted a [PR](https://github.com/circt/llvm/pull/5) to `circt/llvm` (I don't have write access) to fix the CI issue but there are a number of ostensibly unrelated errors, here is a sample: 
Screen Shot 2020-11-22 at 5 17 21 PM

**Comment by lattner** [comment]
It's pretty hard to diagnose that without knowing what is going wrong.  If you need llvm commit access, feel free to request it.  See the llvm developer policy page.

**Comment by mikeurbach** [comment]
@GeorgeLyon I've updated the circt/llvm repo. I believe if you want write access to circt/llvm you will need to be added to the CIRCT org on GitHub, which is different from LLVM commit access. @stephenneuendorffer or @lattner should be able to add you, and then you should be able to do things like fast-forward the master branch to pick up your upstream changes.

I also took the liberty of re-triggering the CI bot, since it will take a bit to rebuild LLVM. I think this may pick up a change to location printing that will cause an FIRParser test to fail. If it does, I have the fix locally and I will push it to this branch.

**Comment by stephenneuendorffer** [comment]
Invite sent to @GeorgeLyon

**Comment by mikeurbach** [comment]
@GeorgeLyon I pushed the update for FIRParser. I think you just need to rebase this on CIRCT master to resolve the other issues.

**Comment by GeorgeLyon** [comment]
@mikeurbach thanks for kicking that off! It looks like it might be error related: https://github.com/llvm/circt/runs/1443724389?check_suite_focus=true#step:8:637

**Comment by GeorgeLyon** [comment]
Whoops, missed your post @mikeurbach, thanks! Looks like @darthscsi updated LLVM this morning but it doesn't seem like he included any of the fixes?

**Comment by mikeurbach** [comment]
I think the update from @darthscsi didn't pick up the change that required the fix I just pushed here, but the update you are making in this PR does.

**Comment by GeorgeLyon** [comment]
Looks like the checks pass now! I'm waiting on another patch I want to get into `llvm-project/master` before I put this up for formal review.

**Comment by mikeurbach** [comment]
@GeorgeLyon FYI I think in the rebasing/force pushing, you dropped the actual change to the LLVM submodule version. I no longer see that being updated in this PR. If you have another version bump coming then I guess it doesn't matter yet.

**Comment by GeorgeLyon** [comment]
@mikeurbach Yes, it went away and not only that, but I couldn't find the original change to cherry-pick. I manually re-added it here: https://github.com/llvm/circt/pull/275/commits/5e5c4794a1798528f25f5823c9c1cbf628152ba4

**Comment by mikeurbach** [comment]
I saw that. What I am trying to point out is this PR no longer updates the version of LLVM.

**Comment by GeorgeLyon** [comment]
Oh right, yeah I missed that too :)

**Comment by GeorgeLyon** [comment]
I've gotten an updated version building, but it seems that many tests now fail with the `'func' op symbol declaration cannot have public visibility`. From what I can gather, this is MLIR becoming more restrictive in what it allows, meaning the ill-formed programs created as error tests now fail earlier than they used to, thus producing a different error message. It looks like @lattner and @seldridge are the authors of the now-failing tests. Can either of you comment as to the correct path forward?

**Comment by mikeurbach** [comment]
That sounds like this: https://llvm.discourse.group/t/psa-funcop-verification-change-for-declarations/2234.

**Comment by GeorgeLyon** [comment]
There is one remaining error in `EmitVerilog/verilog-basic.fir`. Quite a few folks touched that file, so I'm not sure what the best path forward is.

The error output is:
```
<unknown>:0: error: Can't add pass 'LowerFIRRTLToRTL' restricted to 'rtl.module' on a PassManager intended to run on 'firrtl.circuit', did you intend to nest?
```

File: test/firrtl/lower-to-rtl-module.mlir
Comment (mikeurbach) [inline test/firrtl/lower-to-rtl-module.mlir:144]: btw i think you can also use syntax like `func "private" @UnknownFunction ...` now. not sure if there is a preference from the owners here.

Comment (GeorgeLyon) [reply test/firrtl/lower-to-rtl-module.mlir:144]: Oh, I didn't know this. Cool! I'll defer to the owners since I am not really familiar with syntactic sugar in MLIR.

**Comment by mikeurbach** [comment]
@GeorgeLyon about the `Can't add pass ...` error, that sounds like something related to the pass manager requiring nesting to be made explicit now. I fixed this in `firtool` once: https://github.com/llvm/circt/commit/686f2a9773fd71e3d73b4bb91d4177dba241efaa#diff-a76f32ab0202f4b56fb22fb8686da117f6f85a803ab76740ea4a2c5c51c5db55R116.

Maybe the textual pass pipeline is now requiring explicit nesting as well? I think you could update `-pass-pipeline='firrtl.circuit(lower-firrtl-to-rtl)` to `-pass-pipeline='rtl.module(lower-firrtl-to-rtl)` in the verilog-basic.fir test? If it's not that simple, I can help take a look later.

**Comment by GeorgeLyon** [comment]
Looks like that worked on `verilog-basic` though a similar trick did not work on `firrtl/lower-types.mlir` which emits a similar error:
```
<unknown>:0: error: Can't add pass 'LowerFIRRTLTypes' restricted to 'firrtl.module' on a PassManager intended to run on 'firrtl.circuit', did you intend to nest?
```
but changing `-pass-pipeline='firrtl.circuit(lower-firrtl-types)'` to `-pass-pipeline='firrtl.module(lower-firrtl-types)'` doesn't work...

**Comment by mikeurbach** [comment]
Nice! Sorry about `lower-types.mlir`... that one was added by me. I think it would need to be `firrtl.circuit(firrtl.module(lower-firrtl-types))`. It needs to match the nesting of the IR structure, and that's how it is in FIRRTL: modules nest inside circuits.

If that isn't quite right I can take a look... sorry to keep putting it off but I'm trying to wrap something up before recompiling LLVM locally.

**Comment by GeorgeLyon** [comment]
I'm still not completely clear on how passes work, but I'm getting there! And no worries, you've helped me _greatly_ here. Thanks for all the quick replies.

**Comment by GeorgeLyon** [comment]
And it worked,  huzzah!

File: include/circt/Dialect/RTL/Structure.td
Comment (mikeurbach) [inline include/circt/Dialect/RTL/Structure.td:20]: Is this needed for the LLVM update?

File: test/firrtl/lower-to-rtl-module.mlir
Comment (mikeurbach) [reply test/firrtl/lower-to-rtl-module.mlir:144]: Actually seems like the syntax is `func private @UnknownFunction ...` (no quotes around "private"). https://github.com/llvm/llvm-project/blob/b7382ed3fea08da27530a6d6d53f168fc704e4c4/mlir/test/IR/test-func-erase-result.mlir#L5

I would prefer the terser syntax myself. @lattner is the codeowner for all these test cases, and I don't want to speak for him, but I know he expressed interest on the [Discourse discussion](https://llvm.discourse.group/t/rfc-symbol-definition-declaration-x-visibility-checks/2140/12) in having barewords in these scenarios, so that may be the way to go.

**Review by mikeurbach — COMMENTED** [review]
Couple more comments, but this is looking pretty close. Thanks for the iterations @GeorgeLyon.

File: include/circt/Dialect/RTL/Structure.td
Comment (GeorgeLyon) [reply include/circt/Dialect/RTL/Structure.td:20]: Yes, that fixed `error: member 'front' found in multiple base classes of different types`.

**Comment by GeorgeLyon** [comment]
Looks like the terser syntax is unavailable for `firrtl.module`, but since that was just one line I think that is OK

**Comment by mikeurbach** [comment]
@GeorgeLyon sounds good, I guess `func private ...` is available on `FuncOp`, but not all ops implementing `FunctionLike` (like `firrtl.module`).

**Review by mikeurbach — APPROVED** [review]
Looks good to me.

**Comment by lattner** [comment]
Wow, thank you for all the work to land this, I love to see the teamwork!

**Pull request merged**
